resolwe/flow/tests/test_models.py

```diff
@@ -521,6 +524,197 @@ def test_raw_process(self):
         self.assertNotEqual(response.data['id'], self.data.pk)
 
 
+class DuplicateTestCase(TestCase):
+
+    def test_data_duplicate(self):
+        process1 = Process.objects.create(
+            contributor=self.user,
+            type='data:test:first:',
+        )
+        process2 = Process.objects.create(
+            contributor=self.user,
+            input_schema=[
+                {'name': 'data_field', 'type': 'data:test:first:'},
+            ],
+            output_schema=[
+                {'name': 'json_field', 'type': 'basic:json:'},
+            ],
+        )
+
+        data = Data.objects.create(
+            name='Data 1',
+            contributor=self.user,
+            process=process1,
+        )
+        data.status = Data.STATUS_DONE
+        data.save()
+
+        data2 = Data.objects.create(
+            name='Data 2',
+            contributor=self.user,
+            process=process2,
+            started=datetime.now(),
+            input={'data_field': data.id},
+        )
+        data_location = DataLocation.objects.create(subpath='')
+        data_location.subpath = str(data_location.id)
+        data_location.save()
+        data_location.data.add(data2)
+        data2.output = {'json_field': {'foo': 'bar'}}
+        data2.status = Data.STATUS_DONE
+        data2.save()
+        data2.migration_history.create(migration='migration_1')
+        data2.migration_history.create(migration='migration_2')
+
+        # Duplicate.
+        duplicates = Data.objects.filter(id=data2.id).duplicate(self.contributor)
+        self.assertEqual(len(duplicates), 1)
+        duplicate = duplicates[0]
+        self.assertTrue(duplicate.is_duplicate())
+
+        duplicates = Data.objects.filter(id=duplicate.id).duplicate(self.contributor)
+        self.assertEqual(len(duplicates), 1)
+        duplicate_of_duplicate = duplicates[0]
+        self.assertTrue(duplicate_of_duplicate.is_duplicate())
+
+        # Convert original and duplicated to dict.
+        data2_dict = Data.objects.filter(id=data2.id).values()[0]
```

Comment:
You can use:
```
from django.forms.models import model_to_dict
data2_dict = model_to_dict(data2)
```

Reply:
I tried that but there's no `contributor_id` if `model_to_dict` is used.